// @trezor/transport/lib/transports/abstractApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractApiTransport = void 0;
const tslib_1 = require("tslib");
const utils_1 = require("@trezor/utils");
const protocol_1 = require("@trezor/protocol");
const abstract_1 = require("./abstract");
const send_1 = require("../utils/send");
const receive_1 = require("../utils/receive");
const ERRORS = tslib_1.__importStar(require("../errors"));
class AbstractApiTransport extends abstract_1.AbstractTransport {
    constructor({ messages, api, sessionsClient, signal, logger }) {
        super({ messages, signal, logger });
        this.sessionsClient = sessionsClient;
        this.api = api;
    }
    init() {
        return this.scheduleAction(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const handshakeRes = yield this.sessionsClient.handshake();
            this.stopped = !handshakeRes.success;
            return handshakeRes.success
                ? this.success(undefined)
                : this.unknownError('handshake error', []);
        }));
    }
    listen() {
        if (this.listening) {
            return this.error({ error: ERRORS.ALREADY_LISTENING });
        }
        this.listening = true;
        this.api.on('transport-interface-change', descriptors => {
            var _a;
            (_a = this === null || this === void 0 ? void 0 : this.logger) === null || _a === void 0 ? void 0 : _a.debug('new descriptors from api', descriptors);
            this.sessionsClient.enumerateDone({
                descriptors,
            });
        });
        this.sessionsClient.on('descriptors', descriptors => {
            var _a;
            (_a = this === null || this === void 0 ? void 0 : this.logger) === null || _a === void 0 ? void 0 : _a.debug('new descriptors from background', descriptors);
            this.handleDescriptorsChange(descriptors);
        });
        return this.success(undefined);
    }
    enumerate() {
        return this.scheduleAction((signal) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const enumerateResult = yield this.api.enumerate(signal);
            if (!enumerateResult.success) {
                return enumerateResult;
            }
            const descriptors = enumerateResult.payload;
            const enumerateDoneResponse = yield this.sessionsClient.enumerateDone({
                descriptors,
            });
            return this.success(enumerateDoneResponse.payload.descriptors);
        }));
    }
    acquire({ input }) {
        return this.scheduleAction((signal) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { path } = input;
            if (this.listening) {
                this.listenPromise[path] = (0, utils_1.createDeferred)();
            }
            const acquireIntentResponse = yield this.sessionsClient.acquireIntent(input);
            if (!acquireIntentResponse.success) {
                return this.error({ error: acquireIntentResponse.error });
            }
            this.acquiredUnconfirmed[path] = acquireIntentResponse.payload.session;
            const reset = !!input.previous;
            const openDeviceResult = yield this.api.openDevice(path, reset, signal);
            if (!openDeviceResult.success) {
                if (this.listenPromise[path]) {
                    this.listenPromise[path].resolve(openDeviceResult);
                }
                return openDeviceResult;
            }
            this.sessionsClient.acquireDone({ path });
            if (!this.listenPromise[path]) {
                return this.success(acquireIntentResponse.payload.session);
            }
            return this.listenPromise[path].promise.finally(() => {
                delete this.listenPromise[path];
            });
        }), undefined, [ERRORS.DEVICE_DISCONNECTED_DURING_ACTION, ERRORS.SESSION_WRONG_PREVIOUS]);
    }
    release({ path, session, onClose }) {
        return this.scheduleAction(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.listening) {
                this.releaseUnconfirmed[path] = session;
                this.listenPromise[path] = (0, utils_1.createDeferred)();
            }
            const releaseIntentResponse = yield this.sessionsClient.releaseIntent({
                session,
            });
            if (!releaseIntentResponse.success) {
                return this.error({ error: releaseIntentResponse.error });
            }
            const releasePromise = this.releaseDevice(releaseIntentResponse.payload.path);
            if (onClose)
                return this.success(undefined);
            yield releasePromise;
            yield this.sessionsClient.releaseDone({
                path: releaseIntentResponse.payload.path,
            });
            if (!this.listenPromise[path]) {
                return this.success(undefined);
            }
            return this.listenPromise[path].promise
                .then(() => this.success(undefined))
                .finally(() => {
                delete this.listenPromise[path];
            });
        }));
    }
    call({ session, name, data, protocol: customProtocol, }) {
        return this.scheduleAction((signal) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const getPathBySessionResponse = yield this.sessionsClient.getPathBySession({
                session,
            });
            if (!getPathBySessionResponse.success) {
                if (getPathBySessionResponse.error === 'session not found') {
                    return this.error({ error: ERRORS.DEVICE_DISCONNECTED_DURING_ACTION });
                }
                return this.error({ error: ERRORS.UNEXPECTED_ERROR });
            }
            const { path } = getPathBySessionResponse.payload;
            try {
                const protocol = customProtocol || protocol_1.v1;
                const bytes = (0, send_1.buildMessage)({
                    messages: this.messages,
                    name,
                    data,
                    encode: protocol.encode,
                });
                const chunks = (0, send_1.createChunks)(bytes, protocol.getChunkHeader(bytes), this.api.chunkSize);
                const apiWrite = (chunk) => this.api.write(path, chunk, signal);
                const sendResult = yield (0, send_1.sendChunks)(chunks, apiWrite);
                if (!sendResult.success) {
                    throw new Error(sendResult.error);
                }
                const message = yield (0, receive_1.receiveAndParse)(this.messages, () => this.api.read(path, signal).then(result => {
                    if (result.success) {
                        return result.payload;
                    }
                    throw new Error(result.error);
                }), protocol);
                return this.success(message);
            }
            catch (err) {
                if (err.message === ERRORS.DEVICE_DISCONNECTED_DURING_ACTION) {
                    this.enumerate();
                }
                return this.unknownError(err, [
                    ERRORS.DEVICE_DISCONNECTED_DURING_ACTION,
                    ERRORS.DEVICE_NOT_FOUND,
                    ERRORS.INTERFACE_UNABLE_TO_OPEN_DEVICE,
                    ERRORS.INTERFACE_DATA_TRANSFER,
                ]);
            }
        }), { timeout: undefined });
    }
    send({ data, session, name, protocol }) {
        return this.scheduleAction((signal) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const getPathBySessionResponse = yield this.sessionsClient.getPathBySession({
                session,
            });
            if (!getPathBySessionResponse.success) {
                return this.error({ error: getPathBySessionResponse.error });
            }
            const { path } = getPathBySessionResponse.payload;
            try {
                const { encode, getChunkHeader } = protocol || protocol_1.v1;
                const bytes = (0, send_1.buildMessage)({
                    messages: this.messages,
                    name,
                    data,
                    encode,
                });
                const chunks = (0, send_1.createChunks)(bytes, getChunkHeader(bytes), this.api.chunkSize);
                const apiWrite = (chunk) => this.api.write(path, chunk, signal);
                const sendResult = yield (0, send_1.sendChunks)(chunks, apiWrite);
                if (!sendResult.success) {
                    throw new Error(sendResult.error);
                }
                return this.success(undefined);
            }
            catch (err) {
                if (err.message === ERRORS.DEVICE_DISCONNECTED_DURING_ACTION) {
                    this.enumerate();
                }
                return this.unknownError(err, [
                    ERRORS.DEVICE_DISCONNECTED_DURING_ACTION,
                    ERRORS.ABORTED_BY_SIGNAL,
                ]);
            }
        }), { timeout: undefined });
    }
    receive({ session, protocol: customProtocol, }) {
        return this.scheduleAction((signal) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            const getPathBySessionResponse = yield this.sessionsClient.getPathBySession({
                session,
            });
            if (!getPathBySessionResponse.success) {
                return this.error({ error: getPathBySessionResponse.error });
            }
            const { path } = getPathBySessionResponse.payload;
            try {
                const protocol = customProtocol || protocol_1.v1;
                const message = yield (0, receive_1.receiveAndParse)(this.messages, () => this.api.read(path, signal).then(result => {
                    if (!result.success) {
                        throw new Error(result.error);
                    }
                    return result.payload;
                }), protocol);
                return this.success(message);
            }
            catch (err) {
                if (err.message === ERRORS.DEVICE_DISCONNECTED_DURING_ACTION) {
                    this.enumerate();
                }
                return this.unknownError(err, [
                    ERRORS.DEVICE_DISCONNECTED_DURING_ACTION,
                    ERRORS.ABORTED_BY_SIGNAL,
                ]);
            }
        }), { timeout: undefined });
    }
    releaseDevice(path) {
        return this.api.closeDevice(path);
    }
    stop() {
        this.api.on('transport-interface-change', () => {
            var _a;
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.debug('device connected after transport stopped');
        });
        this.stopped = true;
        this.abortController.abort();
    }
}
exports.AbstractApiTransport = AbstractApiTransport;
