// @trezor/transport/lib/transports/udp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UdpTransport = void 0;
const udp_1 = require("../api/udp");
const abstractApi_1 = require("./abstractApi");
const client_1 = require("../sessions/client");
const background_1 = require("../sessions/background");
class UdpTransport extends abstractApi_1.AbstractApiTransport {
    constructor(params) {
        const { messages, logger, signal } = params;
        const sessionsBackground = new background_1.SessionsBackground({ signal });
        const sessionsClient = new client_1.SessionsClient({
            requestFn: args => sessionsBackground.handleMessage(args),
            registerBackgroundCallbacks: () => { },
        });
        sessionsBackground.on('descriptors', descriptors => {
            sessionsClient.emit('descriptors', descriptors);
        });
        super({
            messages,
            api: new udp_1.UdpApi({ logger }),
            logger,
            sessionsClient,
            signal,
        });
        this.name = 'UdpTransport';
    }
    listen() {
        this.api.listen();
        return super.listen();
    }
    stop() {
        if (this.enumerateTimeout) {
            clearTimeout(this.enumerateTimeout);
            this.enumerateTimeout = undefined;
        }
        return super.stop();
    }
}
exports.UdpTransport = UdpTransport;
