// @trezor/transport/lib/api/abstract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractApi = exports.DEVICE_TYPE = void 0;
const utils_1 = require("@trezor/utils");
const result_1 = require("../utils/result");
var DEVICE_TYPE;
(function (DEVICE_TYPE) {
    DEVICE_TYPE[DEVICE_TYPE["TypeT1Hid"] = 0] = "TypeT1Hid";
    DEVICE_TYPE[DEVICE_TYPE["TypeT1Webusb"] = 1] = "TypeT1Webusb";
    DEVICE_TYPE[DEVICE_TYPE["TypeT1WebusbBoot"] = 2] = "TypeT1WebusbBoot";
    DEVICE_TYPE[DEVICE_TYPE["TypeT2"] = 3] = "TypeT2";
    DEVICE_TYPE[DEVICE_TYPE["TypeT2Boot"] = 4] = "TypeT2Boot";
    DEVICE_TYPE[DEVICE_TYPE["TypeEmulator"] = 5] = "TypeEmulator";
})(DEVICE_TYPE || (exports.DEVICE_TYPE = DEVICE_TYPE = {}));
class AbstractApi extends utils_1.TypedEmitter {
    constructor({ logger }) {
        super();
        this.listening = false;
        this.logger = logger;
    }
    success(payload) {
        return (0, result_1.success)(payload);
    }
    error(payload) {
        return (0, result_1.error)(payload);
    }
    unknownError(err, expectedErrors) {
        var _a;
        (_a = this.logger) === null || _a === void 0 ? void 0 : _a.error('transport: abstract api: unknown error', err);
        return (0, result_1.unknownError)(err, expectedErrors);
    }
}
exports.AbstractApi = AbstractApi;
